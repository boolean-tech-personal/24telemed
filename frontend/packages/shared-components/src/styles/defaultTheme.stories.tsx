import type { Meta } from "@storybook/react";
import { capitalize, get } from "lodash-es";
import styled, { css, useTheme } from "styled-components";

import type { PaletteVariants, Theme } from "./theme.types";

interface ColorProps {
  variant: PaletteVariants;
}

function Color({ variant }: ColorProps) {
  const theme = useTheme() as Theme;
  return (
    <ColorRoot variant={variant}>
      <div>
        {capitalize(variant.split(/(\d)./g).join(" ").split(".").join(" "))}
      </div>
      <div>{get(theme.palette, variant)}</div>
    </ColorRoot>
  );
}

const ColorRoot = styled.div<ColorProps>`
  width: 300px;
  height: 137px;
  border-radius: 10px;
  padding: 24px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  ${({ theme }) => css`
    border: 2px solid ${theme.palette.neutral.main};
  `}

  ${({ theme, variant }) => css`
    color: ${get(theme.onPalette, variant)};
    background-color: ${get(theme.palette, variant)};
  `}
`;

function Layout() {
  return (
    <div>
      <Row>
        <Color variant="primary1.main" />
        <Color variant="primary1.light" />
        <Color variant="primary1.lighter" />
      </Row>
      <Row>
        <Color variant="primary2.main" />
        <Color variant="primary2.light" />
        <Color variant="primary2.lighter" />
      </Row>
      <Row>
        <Color variant="secondary1.main" />
        <Color variant="secondary1.light" />
        <Color variant="secondary1.lighter" />
      </Row>
      <Row>
        <Color variant="secondary2.main" />
        <Color variant="secondary2.light" />
        <Color variant="secondary2.lighter" />
      </Row>
      <Row>
        <Color variant="neutral.main" />
        <Color variant="neutral.light" />
        <Color variant="neutral.lighter" />
      </Row>
      <Row>
        <Color variant="common.black" />
        <Color variant="common.white" />
      </Row>
      <Row>
        <Color variant="error" />
        <Color variant="success" />
        <Color variant="notification" />
      </Row>
    </div>
  );
}

const Row = styled.div`
  display: flex;
  gap: 24px;

  &:not(:last-child) {
    margin-bottom: 24px;
  }
`;

const Story: Meta<typeof Layout> = {
  component: Layout,
  title: "Components/Colors",
};
export default Story;

export const Colors = {
  args: {},
};
