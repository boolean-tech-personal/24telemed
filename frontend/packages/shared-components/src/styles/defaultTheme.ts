import type { Breakpoints, PaletteVariants, Theme } from "./theme.types";
import { createThemeBreakpoints } from "./utils";

export const breakpoints: Breakpoints = {
  xs: 600,
  sm: 900,
  md: 1200,
  lg: 1400,
};

const fontFamily = '"Uniform Rounded","Helvetica","Arial",sans-serif';
const fontFamilyHeading =
  '"Uniform Rounded Ex Cond","Helvetica","Arial",sans-serif';

export const colorVariants: Array<PaletteVariants> = [
  "primary1.main",
  "primary1.light",
  "primary1.lighter",
  "primary2.main",
  "primary2.light",
  "primary2.lighter",
  "secondary1.main",
  "secondary1.light",
  "secondary1.lighter",
  "secondary2.main",
  "secondary2.light",
  "secondary2.lighter",
  "secondary3.main",
  "secondary3.light",
  "secondary3.lighter",
  "neutral.main",
  "neutral.light",
  "neutral.lighter",
  "error",
  "success",
  "notification",
  "common.black",
  "common.white",
];

export const defaultTheme: Theme = {
  palette: {
    primary1: {
      main: "#2F4D52",
      light: "#829497",
      lighter: "#D5DBDC",
    },
    primary2: {
      main: "#FFF875",
      light: "#FFFBAC",
      lighter: "#FFFEE3",
    },
    secondary1: {
      main: "#5CD6C2",
      light: "#9DE6DA",
      lighter: "#DEF7F3",
    },
    secondary2: {
      main: "#ED7F75",
      light: "#F19991",
      lighter: "#FBE5E3",
    },
    secondary3: {
      main: "#FFB06C",
      light: "#FFD0A7",
      lighter: "#FFEFE2",
    },
    neutral: {
      main: "#EAEDEE",
      light: "#F2F4F5",
      lighter: "#FBFBFC",
    },
    common: {
      black: "#000000",
      white: "#FFFFFF",
      transparent: "transparent",
    },
    background: {
      main: "#F7F8F8",
      secondary: "#EFF5F6",
      white: "#FFFFFF",
    },
    error: "#FF6363",
    success: "#6FCF97",
    notification: "#9DE6DA",
  },
  onPalette: {
    primary1: {
      main: "#FFFFFF",
      light: "#FFFFFF",
      lighter: "#2F4D52",
    },
    primary2: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    secondary1: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    secondary2: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    secondary3: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    neutral: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    common: {
      black: "#FFFFFF",
      white: "#2F4D52",
      transparent: "#2F4D52",
    },
    background: {
      main: "#2F4D52",
      secondary: "#2F4D52",
      white: "#2F4D52",
    },
    error: "#FFFFFF",
    success: "#FFFFFF",
    notification: "#2F4D52",
  },
  breakpoints: createThemeBreakpoints(breakpoints),
  typography: {
    htmlFontSize: 16,
    fontFamily: fontFamily,
    fontWeight: {
      regular: 400,
      bold: 700,
    },
    defaultColor: "#2F4D52",
    h1: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "3.5rem",
      fontSizeSm: "3rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h2: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "3rem",
      fontSizeSm: "2.25rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h3: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "2.25rem",
      fontSizeSm: "2rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h4: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "2rem",
      fontSizeSm: "1.75rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h5: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "1.75rem",
      fontSizeSm: "1.5rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h6: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "1.5rem",
      fontSizeSm: "1.25rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    bodyLg: {
      fontFamily: fontFamily,
      fontWeight: 400,
      fontSize: "1.125rem",
      lineHeight: 1.5,
    },
    bodyMd: {
      fontFamily: fontFamily,
      fontWeight: 400,
      fontSize: "1rem",
      lineHeight: 1.5,
    },
    bodySm: {
      fontFamily: fontFamily,
      fontWeight: 400,
      fontSize: "0.875rem",
      lineHeight: 1.5,
    },
    bodyXs: {
      fontFamily: fontFamily,
      fontWeight: 400,
      fontSize: "0.75rem",
      lineHeight: 1.5,
    },
    button: {
      fontFamily: fontFamily,
      fontWeight: 700,
      fontSize: "1rem",
      lineHeight: 1.5,
      height: "3.5rem",
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    navigationButton: {
      fontFamily: fontFamily,
      fontWeight: 700,
      fontSize: "1rem",
      lineHeight: 1.5,
      letterSpacing: "-0.02em",
      height: "2.5rem",
      textTransform: "uppercase",
    },
    link: {
      fontWeight: 700,
      textDecoration: "underline",
      fontSize: "1.125rem",
      fontFamily: fontFamily,
    },
  },
  radius: {
    none: "0px",
    progressBar: "4px",
    languageSelect: "8px",
    notification: "8px",
    dropdown: "10px",
    tooltip: "10px",
    textarea: "10px",
    searchResultCard: "10px",
    card: "20px",
    button: "99px",
    toast: "999px",
  },
  border: {
    primary: {
      main: "2px solid #2F4D52",
      light: "1px solid rgba(47, 77, 82, 0.20)",
    },
    secondary: {
      main: "2px dashed #2F4D52",
      light: "2px dashed rgba(47, 77, 82, 0.20)",
    },
    bold: {
      main: "6px solid #2F4D52",
      light: "5px solid #2F4D52",
    },
  },
  spacing: {
    none: "0px",
    xxs: "4px",
    xs: "8px",
    sm: "16px",
    md: "24px",
    lg: "32px",
    xl: "48px",
  },
  height: {
    navBar: "80px",
  },
  sizes: ["xs", "sm", "md", "lg", "xl"],
  shadows: {
    // TODO: discuss also about the shadows, it inherently convey a meaning of elevation depending on how far the shadow is projected
    // I don't think with the current use we have it's working well
    none: "none",
    input: "0px 0px 10px 0px rgba(47,77,82,0.25)",
    md: "8px 13px 30px 0px #2F4D521A",
    lg: "3px 4px 15px 0px #2F4D5240",
    xl: "8px 13px 30px 0px #2F4D5240",
    inset: "3px 4px 10px 0px #2F4D5240 inset",
  },
  icon: {
    fontSizes: {
      sm: "1rem",
      md: "1.5rem",
      lg: "3rem",
    },
  },
  opacity: {
    disabled: 0.3,
    muted: 0.6,
    dimmed: 0.8,
  },
  transitions: {
    easing: {
      easeInOut: "cubic-bezier(0.4, 0, 0.2, 1)",
      easeOut: "cubic-bezier(0.0, 0, 0.2, 1)",
    },
    duration: {
      shortest: 160,
      shorter: 175,
      short: 250,
      standard: 300,
      complex: 375,
      enteringScreen: 225,
      leavingScreen: 195,
    },
  },
  zIndex: {
    appBar: 1200,
    modal: 1500,
    toaster: 1600,
    selectPopper: 1700,
    tooltip: 1800,
  },
  button: {
    primary: {
      backgroundColor: "#2F4D52",
      color: "#FFFFFF",
      hoverBackgroundColor: "#D5DBDC",
      disabledBackgroundColor: "#D5DBDC",
      disabledColor: "#829497",
      hoverColor: "#FFFFFF",
      focusBackgroundColor: "#2F4D52",
      pressedBackgroundColor: "#829497",
    },
    secondary: {
      backgroundColor: "#EAEDEE",
      color: "#2F4D52",
      hoverBackgroundColor: "#EFF5F6",
      disabledBackgroundColor: "#D5DBDC",
      disabledColor: "#829497",
      hoverColor: "##FFFFFF",
      focusBackgroundColor: "#EAEDEE",
      pressedBackgroundColor: "#829497",
    },
    tertiary: {
      backgroundColor: "transparent",
      color: "#2F4D52",
      hoverBackgroundColor: "#EFF5F6",
      disabledBackgroundColor: "transparent",
      disabledColor: "#829497",
      hoverColor: "#2F4D52",
      focusBackgroundColor: "transparent",
      pressedBackgroundColor: "#D5DBDC",
    },
  },
};
