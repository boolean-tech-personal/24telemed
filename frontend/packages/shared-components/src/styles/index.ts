export { defaultTheme, colorVariants, breakpoints } from "./defaultTheme";
export * from "./theme.types";
export * from "./utils";
export * from "./createTheme";
export * from "./useBreakpoints";
