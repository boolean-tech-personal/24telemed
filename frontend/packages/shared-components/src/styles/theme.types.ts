import type { ThemedCssFunction } from "styled-components";
import { RecursiveKeyOf } from "../utils.types";



export type Sizes = "xxs" | "xs" | "sm" | "md" | "lg" | "xl";
export type OptionalSizes = Sizes | "none";

export type FromSize<T extends OptionalSizes, V> = {
  [K in T]: V;
};

export interface Tones {
  main: string;
  light: string;
  lighter: string;
}

export interface MainPalettes {
  primary1: Tones;
  primary2: Tones;
  secondary1: Tones;
  secondary2: Tones;
  secondary3: Tones;
  neutral: Tones;
}

export interface Palettes extends MainPalettes {
  error: string;
  success: string;
  notification: string;
  common: {
    black: string;
    white: string;
    transparent: string;
    warning: string;
  };
  background: {
    main: string;
    secondary: string;
    white: string;
  };
}

export type BreakpointSizes = "xs" | "sm" | "md" | "lg";

export type Breakpoints = FromSize<BreakpointSizes, number>;

export interface Breakpoint {
  value: number;
  up: MediaWrapperFunction;
  down: MediaWrapperFunction;
}

export type ThemeBreakpoints = {
  [key in keyof Breakpoints]: Breakpoint;
};

export interface TypographyVariant {
  fontFamily: string;
  fontWeight: number;
  fontSize: string;
  lineHeight: number;
  letterSpacing?: string;
}

export interface HeadingVariant extends TypographyVariant {
  fontSizeSm: string;
  textTransform: string;
}

export interface ButtonVariant extends TypographyVariant {
  height: string;
  textTransform: string;
}

export interface TypographyLink {
  fontWeight: number;
  textDecoration: string;
  fontSize: string;
  fontFamily: string;
}

export interface FontWeight {
  regular: number;
  bold: number;
}

export type FontWeightVariants = keyof FontWeight;

export interface TypographyTheme {
  htmlFontSize: number;
  fontFamily: string;
  fontWeight: FontWeight;
  defaultColor: string;
  h1: HeadingVariant;
  h2: HeadingVariant;
  h3: HeadingVariant;
  h4: HeadingVariant;
  h5: HeadingVariant;
  h6: HeadingVariant;
  bodyLg: TypographyVariant;
  bodyMd: TypographyVariant;
  bodySm: TypographyVariant;
  bodyXs: TypographyVariant;
  button: ButtonVariant;
  navigationButton: ButtonVariant;
  link: TypographyLink;
}

export type TypographyVariants =
  | "h1"
  | "h2"
  | "h3"
  | "h4"
  | "h5"
  | "h6"
  | "bodyLg"
  | "bodyMd"
  | "bodySm"
  | "bodyXs";

export type Radius = {
  none: string;
  progressBar: string;
  languageSelect: string;
  notification: string;
  dropdown: string;
  tooltip: string;
  textarea: string;
  searchResultCard: string;
  card: string;
  button: string;
  toast: string;
};

export type RadiusVariants = keyof Radius;

export type Border = {
  primary: {
    main: string;
    light: string;
  };
  secondary: {
    main: string;
    light: string;
  };
  bold: {
    main: string;
    light: string;
  };
};

export type Spacing = FromSize<
  "none" | "xxs" | "xs" | "sm" | "md" | "lg" | "xl",
  string
>;

export type SpacingVariants = keyof Spacing;

export type Height = {
  navBar: string;
};

export type Shadows = FromSize<"none" | "md" | "lg" | "xl", string> & {
  input: string;
  inset: string;
};

export type ShadowVariants = keyof Shadows;

export type IconFontSize = FromSize<"sm" | "md" | "lg", string>;

export type IconFontSizeVariants = keyof IconFontSize;

export interface Icon {
  fontSizes: IconFontSize;
}

export interface Opacity {
  disabled: 0.3;
  muted: 0.6;
  dimmed: 0.8;
}

export interface Easing {
  easeInOut: string;
  easeOut: string;
}

export type EasingVariants = keyof Easing;

export interface Durations {
  shortest: number;
  shorter: number;
  short: number;
  standard: number;
  complex: number;
  enteringScreen: number;
  leavingScreen: number;
}

export type DurationVariants = keyof Durations;

export interface Transitions {
  easing: Easing;
  duration: Durations;
}

export interface ZIndex {
  toaster: number;
  appBar: number;
  modal: number;
  selectPopper: number;
  tooltip: number;
}

export interface Theme {
  palette: Palettes;
  onPalette: Palettes;
  breakpoints: ThemeBreakpoints;
  typography: TypographyTheme;
  radius: Radius;
  border: Border;
  spacing: Spacing;
  height: Height;
  sizes: Array<Sizes>;
  shadows: Shadows;
  icon: Icon;
  opacity: Opacity;
  transitions: Transitions;
  zIndex: ZIndex;
  button: {
    primary: {
      backgroundColor: string;
      color: string;
      hoverBackgroundColor: string;
      disabledBackgroundColor: string;
      disabledColor: string;
      hoverColor: string;
      focusBackgroundColor: string;
      pressedBackgroundColor: string;
      
    };
    secondary: {
      backgroundColor: string;
      color: string;
      hoverBackgroundColor: string;
      disabledBackgroundColor: string;
      disabledColor: string;
      hoverColor: string;
      focusBackgroundColor: string;
      pressedBackgroundColor: string;
    };
    tertiary: {
      backgroundColor: string;
      color: string;
      hoverBackgroundColor: string;
      disabledBackgroundColor: string;
      disabledColor: string;
      hoverColor: string;
      focusBackgroundColor: string;
      pressedBackgroundColor: string;
    };
  };

}

export type PaletteVariants = RecursiveKeyOf<Palettes>;

export type MainPaletteVariants = keyof MainPalettes;

type ThemedCssFunctionReturn = ReturnType<ThemedCssFunction<Theme>>;

export type MediaWrapperFunction = (
  cssResult: ThemedCssFunctionReturn,
) => ThemedCssFunctionReturn;


export type NewButtonVariant = "primary" | "secondary" | "tertiary";