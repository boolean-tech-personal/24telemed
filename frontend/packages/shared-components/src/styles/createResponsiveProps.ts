import { css } from "styled-components";

import { breakpoints } from "./defaultTheme";
import { Breakpoints, Theme } from "./theme.types";

export type ResponsivePropsBase = Record<string, unknown>;

export type ResponsiveProps<TResponsivePropsBase extends ResponsivePropsBase> =
  {
    [K in keyof TResponsivePropsBase as `${keyof Breakpoints}${Capitalize<
      string & K
    >}`]?: TResponsivePropsBase[K];
  };

export type PropsCssMap<TResponsivePropsBase extends ResponsivePropsBase> =
  Required<{
    [K in keyof TResponsivePropsBase as Capitalize<string & K>]: string;
  }>;

export type ResponsivePropsGetters<
  TResponsivePropsBase extends ResponsivePropsBase,
> = Partial<{
  [K in keyof PropsCssMap<TResponsivePropsBase>]: (
    theme: Theme,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    value: any,
  ) => string;
}>;

export const createResponsiveProps = <
  TResponsivePropsBase extends ResponsivePropsBase,
>(
  propsCssMap: PropsCssMap<TResponsivePropsBase>,
  responsivePropsGetters?: ResponsivePropsGetters<TResponsivePropsBase>,
) =>
  function responsiveProps(
    props: ResponsiveProps<TResponsivePropsBase> & { theme: Theme },
  ) {
    const breakpointKeys = Object.keys(breakpoints) as Array<keyof Breakpoints>;
    const cssMapEntries = Object.entries(propsCssMap);

    const styles = [];
    for (let i = 0; i < breakpointKeys.length; i++) {
      const breakpoint = breakpointKeys[i];
      let cssProperties = "";

      for (let j = 0; j < cssMapEntries.length; j++) {
        const [propBase, cssProp] = cssMapEntries[j] as [
          keyof PropsCssMap<TResponsivePropsBase>,
          string,
        ];
        if (typeof propBase !== "string") continue;

        const prop =
          `${breakpoint}${propBase}` as keyof ResponsiveProps<TResponsivePropsBase>;
        const value = props[prop];

        if (value !== undefined) {
          const transformedValue = responsivePropsGetters?.[propBase]?.(
            props.theme,
            value,
          );
          cssProperties += `${cssProp}: ${transformedValue ?? value};`;
        }
      }

      if (cssProperties) {
        styles.push(
          props.theme.breakpoints[breakpoint].down(css`
            ${cssProperties}
          `),
        );
      }
    }

    return styles;
  };
