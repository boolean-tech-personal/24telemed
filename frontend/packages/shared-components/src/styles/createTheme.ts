import defaultsDeep from "lodash/defaultsDeep";
import type { PartialDeep } from "type-fest";

import { defaultTheme } from "./defaultTheme";
import type { Theme } from "./theme.types";

export function createTheme(themeOverride: PartialDeep<Theme> = {}): Theme {
  return defaultsDeep(themeOverride, defaultTheme);
}
