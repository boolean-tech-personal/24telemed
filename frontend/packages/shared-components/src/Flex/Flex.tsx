import styled, { css } from "styled-components";

import { SpacingVariants } from "../styles";
import {
  ResponsiveProps,
  ResponsivePropsBase,
  createResponsiveProps,
} from "../styles/createResponsiveProps";
import { FC } from "react";

export type Justify =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-around"
  | "space-between"
  | "space-evenly";

type Align = "flex-start" | "flex-end" | "center";

interface FlexResponsivePropsBase extends ResponsivePropsBase {
  direction?: "column" | "row";
  gap?: SpacingVariants;
  justify?: Justify;
  align?: Align;
  padding?: SpacingVariants;
}

export interface FlexProps
  extends FlexResponsivePropsBase,
    ResponsiveProps<FlexResponsivePropsBase> {
  marginBottom?: SpacingVariants;
  className?: string;
  fullWidth?: boolean;
  fullHeight?: boolean;
}

const responsiveProps = createResponsiveProps<FlexResponsivePropsBase>(
  {
    Direction: "flex-direction",
    Gap: "gap",
    Justify: "justify-content",
    Align: "align-items",
    Padding: "padding",
  },
  {
    Gap: (theme, value: SpacingVariants) => theme.spacing[value],
    Padding: (theme, value: SpacingVariants) => theme.spacing[value],
  },
);

export const Flex: FC<FlexProps> = styled.div<FlexProps>`
  display: flex;
  flex-direction: ${({ direction = "row" }) => direction};
  justify-content: ${({ justify = "flex-start" }) => justify};
  align-items: ${({ align, direction }) =>
    align ?? (direction === "column" ? "stretch" : "center")};
  gap: ${({ theme, gap = "sm" }) => theme.spacing[gap]};
  margin-bottom: ${({ theme, marginBottom = "none" }) =>
    theme.spacing[marginBottom]};
  ${({ theme, padding }) =>
    padding &&
    css`
      padding: ${theme.spacing[padding]};
    `};
  ${({ fullWidth }) =>
    fullWidth &&
    css`
      width: 100%;
    `};
  ${({ fullHeight }) =>
    fullHeight &&
    css`
      height: 100%;
    `};

  ${responsiveProps}
`;