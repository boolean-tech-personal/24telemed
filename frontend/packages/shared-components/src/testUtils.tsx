import React, { ReactElement } from "react";

import "@testing-library/jest-dom/vitest";
import { RenderOptions, render } from "@testing-library/react";
import "jest-styled-components";
import { ThemeProvider } from "styled-components";
import { vi } from "vitest";

import { defaultTheme } from "./styles/defaultTheme";

export * from "./testHelpers";

const AllTheProviders = ({ children }: { children: ReactElement }) => {
  return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
  // @ts-ignore
) => render(ui, { wrapper: AllTheProviders, ...options });

const matchMedia = (mediaRule: string) =>
  vi.fn((media: string) => ({
    matches: media === mediaRule,
    media,
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
  }));

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render, matchMedia };
