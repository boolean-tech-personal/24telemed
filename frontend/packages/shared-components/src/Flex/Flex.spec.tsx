import { defaultTheme } from "../styles";
import { mediaQueries, render } from "../testUtils";
import { Flex } from "./Flex";

describe("Flex", () => {
  it("should render with default props", () => {
    const { container } = render(<Flex />);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveStyleRule("display", "flex");
    expect(container.firstChild).toHaveStyleRule("flex-direction", "row");
    expect(container.firstChild).toHaveStyleRule(
      "justify-content",
      "flex-start",
    );
    expect(container.firstChild).toHaveStyleRule("align-items", "center");
  });

  it("should apply responsive props correctly", () => {
    const { container } = render(<Flex smDirection="column" />);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveStyleRule("flex-direction", "column", {
      media: mediaQueries.sm.down,
    });
  });

  it("should apply responsive gap prop correctly", () => {
    const { container } = render(<Flex smGap="sm" />);
    expect(container.firstChild).toHaveStyleRule("gap", "16px", {
      media: mediaQueries.sm.down,
    });
  });

  it("should apply marginBottom prop", () => {
    const { container } = render(<Flex marginBottom="lg" />);
    expect(container.firstChild).toHaveStyleRule(
      "margin-bottom",
      defaultTheme.spacing.lg,
    );
  });

  it("should apply justify and align props", () => {
    const { container } = render(<Flex justify="center" align="flex-end" />);
    expect(container.firstChild).toHaveStyleRule("justify-content", "center");
    expect(container.firstChild).toHaveStyleRule("align-items", "flex-end");
  });

  it("should override align-items when direction is column", () => {
    const { container } = render(<Flex direction="column" />);
    expect(container.firstChild).toHaveStyleRule("align-items", "stretch");
  });
});
