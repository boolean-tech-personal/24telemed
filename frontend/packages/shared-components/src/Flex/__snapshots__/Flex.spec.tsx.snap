// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Flex > should apply responsive props correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 16px;
  margin-bottom: 0px;
}

@media (max-width:899.95px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
  class="c0"
/>
`;

exports[`Flex > should render with default props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 16px;
  margin-bottom: 0px;
}

<div
  class="c0"
/>
`;
